compparallel.go

```diff
@@ -326,115 +325,131 @@ func getChannelOrErrorParallel[T any](
-				" and ignore errors ", r.IgnoreError,
-			)
-			tim := time.NewTimer(r.ExecuteAfter)
-			defer tim.Stop()
-			select {
-			case <-ctx.Done():
-				log.Debug("getChannelOrErrorParallel: stopping execution on router on context done for router ", r.Router,
-					" with timeout ", r.Timeout,
-					" and ignore errors ", r.IgnoreError,
-				)
-				return
-			case <-tim.C:
-				ctx, cancel := withCancelAndOptionalTimeout(ctx, r.Timeout)
-				defer cancel()
+			defer fwg.Done()
+			defer func() {
+				var remainingBlockers uint64
+				if r.DoNotWaitForStreamingResponses {
+					remainingBlockers = blocking.Load()
+				} else {
+					var minusOne uint64
+					minusOne--
+					remainingBlockers = blocking.Add(minusOne)
+				}
 
-				log.Debug("getChannelOrErrorParallel: calling router function for router ", r.Router,
-					" with timeout ", r.Timeout,
-					" and ignore errors ", r.IgnoreError,
-				)
-				valueChan, err := f(ctx, r.Router)
-				if err != nil && !r.IgnoreError {
-					log.Debug("getChannelOrErrorParallel: error calling router function for router ", r.Router,
-						" with timeout ", r.Timeout,
-						" and ignore errors ", r.IgnoreError,
-						" with error ", err,
-					)
-					select {
-					case <-ctx.Done():
-					case errCh <- err:
-					}
+				if remainingBlockers == 0 && sent.Load() {
+					cancelAll()
+				}
+			}()
+
+			if r.ExecuteAfter != 0 {
+				tim := time.NewTimer(r.ExecuteAfter)
+				defer tim.Stop()
+				select {
+				case <-ctx.Done():
 					return
+				case <-tim.C:
+					// ready
 				}
-				for {
-					select {
-					case <-ctx.Done():
-						log.Debug("getChannelOrErrorParallel: stopping execution on router on context done inside loop for router ", r.Router,
-							" with timeout ", r.Timeout,
-							" and ignore errors ", r.IgnoreError,
-						)
+			}
+
+			ctx, cancel := withCancelAndOptionalTimeout(ctx, r.Timeout)
+			defer cancel()
+
+			valueChan, err := f(ctx, r.Router)
+			if err != nil {
+				if r.IgnoreError {
+					return
+				}
+
+				errorsLk.Lock()
+				defer errorsLk.Unlock()
+				if errors == nil {
+					return
+				}
+				errors = append(errors, err)
+
+				return
+			}
+
+			for first := true; true; first = false {
+				select {
+				case <-ctx.Done():
+					return
+				case val, ok := <-valueChan:
+					if !ok {
 						return
-					case val, ok := <-valueChan:
-						log.Debug("getChannelOrErrorParallel: getting channel value for router ", r.Router,
-							" with timeout ", r.Timeout,
-							" and ignore errors ", r.IgnoreError,
-							" closed channel: ", ok,
-						)
-						if !ok {
-							return
-						}
+					}
 
-						select {
-						case <-ctx.Done():
-							log.Debug("getChannelOrErrorParallel: stopping execution on router on context done inside select for router ", r.Router,
-								" with timeout ", r.Timeout,
-								" and ignore errors ", r.IgnoreError,
-							)
-							return
-						case outCh <- val:
+					if first {
+						errorsLk.Lock()
```

Comment:
if `errorsLk` is protecting `outCh` too that at least needs a comment in not a name change